storagenode/gracefulexit/worker.go

```
@@ -4,48 +4,248 @@
+
+				continue
+			}
+
+			err = verifyPieceHash(ctx, addrLimit.GetLimit(), pieceHash, originalHash.Hash)
+			if err != nil {
+				worker.log.Error("failed hash verification.", zap.Stringer("satellite ID", satelliteID), zap.Stringer("piece ID", pieceID), zap.Error(errs.Wrap(err)))
+				worker.handleFailure(ctx, pb.TransferFailed_HASH_VERIFICATION, pieceID, satelliteID, c.Send)
+
+				continue
+			}
+
+			success := &pb.StorageNodeMessage{
+				Message: &pb.StorageNodeMessage_Succeeded{
+					Succeeded: &pb.TransferSucceeded{
+						OriginalPieceId:      msg.TransferPiece.OriginalPieceId,
+						OriginalPieceHash:    &pb.PieceHash{PieceId: msg.TransferPiece.OriginalPieceId},
+						OriginalOrderLimit:   &originalOrderLimit,
+						ReplacementPieceHash: pieceHash,
+					},
+				},
+			}
+			err = c.Send(success)
+			if err != nil {
+				return errs.Wrap(err)
+			}
+		case *pb.SatelliteMessage_DeletePiece:
+			pieceID := msg.DeletePiece.OriginalPieceId
+			err := worker.store.Delete(ctx, satelliteID, pieceID)
+			if err != nil {
+				worker.log.Error("failed to delete piece.", zap.Stringer("satellite ID", satelliteID), zap.Stringer("piece ID", pieceID), zap.Error(errs.Wrap(err)))
+			}
+		case *pb.SatelliteMessage_ExitFailed:
+			worker.log.Error("graceful exit failed.", zap.Stringer("satellite ID", satelliteID), zap.Stringer("reason", msg.ExitFailed.Reason))
+
+			err = worker.satelliteDB.CompleteGracefulExit(ctx, satelliteID, time.Now(), satellites.ExitFailed, msg.ExitFailed.GetExitFailureSignature())
+			if err != nil {
+				return errs.Wrap(err)
+			}
+			break
+		case *pb.SatelliteMessage_ExitCompleted:
+			worker.log.Info("graceful exit completed.", zap.Stringer("satellite ID", satelliteID))
+
+			err = worker.satelliteDB.CompleteGracefulExit(ctx, satelliteID, time.Now(), satellites.ExitSucceeded, msg.ExitCompleted.GetExitCompleteSignature())
+			if err != nil {
+				return errs.Wrap(err)
+			}
+			break
+		default:
+			// TODO handle err
```

Should we return a custom error in this case? This should never happen.